diff -urNp hal-0.5.14.old/hald/haldaemon.in hal-0.5.14/hald/haldaemon.in
--- hal-0.5.14.old/hald/haldaemon.in	2011-02-01 17:22:19.049150295 +0000
+++ hal-0.5.14/hald/haldaemon.in	2011-02-01 17:25:16.305914252 +0000
@@ -18,6 +18,9 @@
 processname=hald
 servicename=haldaemon
 
+if [ -f @sysconfdir@/sysconfig/haldaemon ]; then
+    HALD_ARGS=`cat @sysconfdir@/sysconfig/haldaemon`
+fi
 RETVAL=0
 
 #
@@ -39,7 +42,7 @@ start() {
 	# Check if it is already running
 	if [ ! -f @localstatedir@/lock/subsys/$servicename ]; then
 		echo -n $"Starting HAL daemon: "
-		daemon --check $servicename $processname
+		daemon --check $servicename $processname $HALD_ARGS
 		RETVAL=$?
 		[ $RETVAL -eq 0 ] && touch @localstatedir@/lock/subsys/$servicename
 		echo

