diff -urNp hal-0.5.14.old/tools/hal-device.c hal-0.5.14/tools/hal-device.c
--- hal-0.5.14.old/tools/hal-device.c	2011-04-01 09:29:06.000000000 +0100
+++ hal-0.5.14/tools/hal-device.c	2011-04-01 09:32:50.000000000 +0100
@@ -193,6 +193,7 @@ int dump_devices(LibHalContext *hal_ctx,
 	char **device_names;
 	DBusError error;
 	char *udi = NULL;
+	dbus_bool_t udi_exists;
 
 	if (arg) {
 		if (*arg == '/') {
@@ -218,6 +219,19 @@ int dump_devices(LibHalContext *hal_ctx,
 			return 31;
 		}
 	} else {
+
+		/* check UDI exists */
+		udi_exists = libhal_device_exists (hal_ctx, udi, &error);
+		if (!udi_exists) {
+			fprintf (stderr, "error: UDI %s does not exist\n", udi);
+			return 22;
+		}
+		if (dbus_error_is_set(&error)) {
+			fprintf (stderr, "error: libhal_device_exists: %s: %s\n", error.name, error.message);
+			LIBHAL_FREE_DBUS_ERROR (&error);
+			return 1;
+		}
+
 		device_names = calloc(2, sizeof *device_names);
 		device_names[0] = strdup(udi);
 		num_devices = 1;
diff -urNp hal-0.5.14.old/tools/hal_set_property.c hal-0.5.14/tools/hal_set_property.c
--- hal-0.5.14.old/tools/hal_set_property.c	2011-04-01 09:29:06.000000000 +0100
+++ hal-0.5.14/tools/hal_set_property.c	2011-04-01 09:31:49.000000000 +0100
@@ -104,6 +104,7 @@ int
 main (int argc, char *argv[])
 {
 	dbus_bool_t rc = 0;
+	dbus_bool_t udi_exists;
 	char *udi = NULL;
 	char *key = NULL;
 	char *str_value = NULL;
@@ -248,6 +249,18 @@ main (int argc, char *argv[])
 		}
 	}
 
+	/* check UDI exists */
+	udi_exists = libhal_device_exists (hal_ctx, udi, &error);
+	if (!udi_exists) {
+		fprintf (stderr, "error: UDI %s does not exist\n", udi);
+		return 1;
+	}
+	if (dbus_error_is_set(&error)) {
+		fprintf (stderr, "error: libhal_device_exists: %s: %s\n", error.name, error.message);
+		LIBHAL_FREE_DBUS_ERROR (&error);
+		return 1;
+	}
+
 	if (remove) {
 		rc = libhal_device_remove_property (hal_ctx, udi, key, &error);
 		if (!rc) {
