diff --git a/fdi/policy/10osvendor/10-x11-input.fdi b/fdi/policy/10osvendor/10-x11-input.fdi
index 752c4ab..48633d9 100644
--- a/fdi/policy/10osvendor/10-x11-input.fdi
+++ b/fdi/policy/10osvendor/10-x11-input.fdi
@@ -29,6 +29,13 @@
       </match>
     </match>
 
+    <match key="info.capabilities" contains="input.touchscreen">
+      <match key="/org/freedesktop/Hal/devices/computer:system.kernel.name"
+             string="Linux">
+        <merge key="input.x11_driver" type="string">evdev</merge>
+      </match>
+    </match>
+
     <match key="info.capabilities" contains="input.keys">
       <!-- If we're using Linux, we use evdev by default (falling back to
            keyboard otherwise). -->
diff --git a/hald/linux/device.c b/hald/linux/device.c
index e7c9d4f..d07fa9b 100644
--- a/hald/linux/device.c
+++ b/hald/linux/device.c
@@ -1180,6 +1180,11 @@ input_test_abs (HalDevice *d, const char *sysfs_path)
 				hal_device_add_capability (d, "input.mouse");
 				goto out;
 			}
+
+			if (test_bit (BTN_TOUCH, bitmask_key)) {
+				hal_device_add_capability (d, "input.touchscreen");
+				goto out;
+			}
 		}
 	}
 out:
